ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/trielog/TrieLogPruner.java

```diff
@@ -0,0 +1,188 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.hyperledger.besu.ethereum.bonsai.trielog;
+
+import org.hyperledger.besu.datatypes.Hash;
+import org.hyperledger.besu.ethereum.bonsai.storage.BonsaiWorldStateKeyValueStorage;
+import org.hyperledger.besu.ethereum.chain.Blockchain;
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+import org.hyperledger.besu.ethereum.core.ProcessableBlockHeader;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Stream;
+
+import com.google.common.collect.Multimap;
+import com.google.common.collect.TreeMultimap;
+import org.apache.tuweni.bytes.Bytes32;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TrieLogPruner {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TrieLogPruner.class);
+
+  public static final long DEFAULT_RETENTION_THRESHOLD = 512L;
+  public static final long MINIMUM_RETENTION_THRESHOLD = DEFAULT_RETENTION_THRESHOLD;
+  public static int DEFAULT_PRUNING_LIMIT = 30_000;
+
+  private final int pruningLimit;
+  private final int loadingLimit;
+  private final BonsaiWorldStateKeyValueStorage rootWorldStateStorage;
+  private final Blockchain blockchain;
+  private final long numBlocksToRetain;
+
+  private static final Multimap<Long, Hash> trieLogBlocksAndForksByDescendingBlockNumber =
+      TreeMultimap.create(Comparator.reverseOrder(), Comparator.naturalOrder());
+
+  public TrieLogPruner(
+      final BonsaiWorldStateKeyValueStorage rootWorldStateStorage,
+      final Blockchain blockchain,
+      final long numBlocksToRetain,
+      final int pruningLimit) {
+    this.rootWorldStateStorage = rootWorldStateStorage;
+    this.blockchain = blockchain;
+    this.numBlocksToRetain = numBlocksToRetain;
+    this.pruningLimit = pruningLimit;
+    this.loadingLimit = pruningLimit; // same as pruningLimit for now
+  }
+
+  public void initialize() {
+    preloadCache();
+  }
+
+  private void preloadCache() {
+    LOG.atInfo()
+        .setMessage("Loading first {} trie logs from database...")
+        .addArgument(loadingLimit)
+        .log();
+    final Stream<byte[]> trieLogKeys = rootWorldStateStorage.streamTrieLogKeys(loadingLimit);
+    final AtomicLong count = new AtomicLong();
+    trieLogKeys.forEach(
+        blockHashAsBytes -> {
+          Hash blockHash = Hash.wrap(Bytes32.wrap(blockHashAsBytes));
+          final Optional<BlockHeader> header = blockchain.getBlockHeader(blockHash);
+          if (header.isPresent()) {
+            trieLogBlocksAndForksByDescendingBlockNumber.put(header.get().getNumber(), blockHash);
```

Comment:
Is it possible that not enough trielog are not loaded such that we don't fill the gap between head and minimum retention threshold? I think it could only happen if there are a very large number of trie logs as the pruning limit in most cases would load all the tries.

Reply:
As discussed, there will always be future opportunities to prune the gap each time besu is started (and the cache is preloaded). This works because the pruningLimit >> retainedTrieLogs. It might be worth enforcing this property in code to make sure the trie logs can't end up growing again.